codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@Scottmitch I agree that those are the steps that *should* happen.  What I'm saying is that from the perspective of the flow controller, it shouldn't trust that the encoder behaves in this way ... doing so would make the interaction between the flow controller and encoder fragile.  The flow controller should also be able to handle this sequence:

1. Flow controller invokes payload.write().
2. "Encoder" (FlowControlled class that lives in the encoder) writes data.
3. Repeat 1 and 2 until no more to write
6. "Encoder" changes state via lifecycle manager if EOS.
4. "Encoder" tells listener (which would be the flow controller) that it is done writing.
5. Flow controller cleans up its queue

I kept your numbers for clarity, but the sequence would be in the order they appear.

So by "protect", I mean that the flow controller's queue should not be adversely affected by a slightly misbehaving encoder.  I'm not sold on the idea of popping the queue either ... maybe there's a better way to handle this?  Maybe we can somehow be aware that we're in the middle of a `write` callback and postpone clearing the queue until after it completes? WDYT?